src/logger.cr

```diff
@@ -1,178 +1,56 @@
-# and would set the `Logger` to `DEBUG`.
-#
-# If logging to multiple locations is required, an `IO::MultiWriter` can be
-# used.
-#
-# ### Example
-#
-# ```
-# require "logger"
-#
-# log = Logger.new(STDOUT)
-# log.level = Logger::WARN
-#
-# # or
-# log = Logger.new(STDOUT, level: Logger::WARN)
-#
-# log.debug("Created logger")
-# log.info("Program started")
-# log.warn("Nothing to do!")
-#
-# begin
-#   File.each_line("/foo/bar.log") do |line|
-#     unless line =~ /^(\w+) = (.*)$/
-#       log.error("Line in wrong format: #{line}")
-#     end
-#   end
-# rescue err
-#   log.fatal("Caught exception; exiting")
-#   log.fatal(err)
-# end
-# ```
-class Logger
-  property level : Severity
-  property progname : String
-
-  # Customizable `Proc` (with a reasonable default)
-  # which the `Logger` uses to format and print its entries.
-  #
-  # Use this setter to provide a custom formatter.
-  # The `Logger` will invoke it with the following arguments:
-  #  - severity: a `Logger::Severity`
-  #  - datetime: `Time`, the entry's timestamp
-  #  - progname: `String`, the program name, if set when the logger was built
-  #  - message: `String`, the body of a message
-  #  - io: `IO`, the Logger's stream, to which you must write the final output
-  #
-  # Example:
-  #
-  # ```
-  # require "logger"
-  #
-  # logger = Logger.new(STDOUT)
-  # logger.progname = "YodaBot"
-  #
-  # logger.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
-  #   label = severity.unknown? ? "ANY" : severity.to_s
-  #   io << label[0] << ", [" << datetime << " #" << Process.pid << "] "
-  #   io << label.rjust(5) << " -- " << progname << ": " << message
-  # end
-  #
-  # logger.warn("Fear leads to anger. Anger leads to hate. Hate leads to suffering.")
-  #
-  # # Prints to the console:
-  # # "W, [2017-05-06 18:00:41 -0300 #11927]  WARN --
-  # #  YodaBot: Fear leads to anger. Anger leads to hate. Hate leads to suffering."
-  # ```
-  property formatter
-
-  # A logger severity level.
-  enum Severity
-    # Low-level information for developers
-    DEBUG
-
-    # Generic (useful) information about system operation
-    INFO
-
-    # A warning
-    WARN
-
-    # A handleable error condition
-    ERROR
-
-    # An unhandleable error that results in a program crash
-    FATAL
+require "logger/entry"
+require "logger/filter"
+require "logger/emitter"
+require "logger/severity"
 
-    UNKNOWN
-  end
+class Logger
+  alias FilterType = Filter | Severity | (Entry -> Bool)
+  alias EmitterType = Emitter | (Entry -> Nil)
+
+  module Base
+    abstract def component : String
+    abstract def filter : FilterType?
+    abstract def emitter : EmitterType?
+
+    def log(entry : Entry)
+      case ff = filter
+      when Filter, Proc(Entry, Bool)
+        return unless ff.call(entry)
+      when Severity
+        return unless entry.severity >= ff
+      end
+      emitter.try &.call(entry)
```

Comment:
It was hard to follow how the `Logger.class` that is a `Logger::Base` ends up chained as the default `emitter`.